src/source/Ice/IceAgentStateMachine.c

```diff
@@ -732,8 +732,8 @@ STATUS executeDisconnectedIceAgentState(UINT64 customData, UINT64 time)
         pIceAgent->iceAgentCallbacks.connectionStateChangedFn(pIceAgent->customData, ICE_AGENT_STATE_DISCONNECTED);
     }
 
-    // step out of disconnection state to retry
-    CHK_STATUS(stepIceAgentStateMachine(pIceAgent));
+    // step out of disconnection state to retry. Do not use stepIceAgentState machine because lock is not re-entrant.
```

Comment:
Re-wording the comment - perhaps this means that this might be run from another thread that already holds the lock?

Reply:
BTW, could this be easier solved by making the lock re-entrant?

Reply:
I purposely made the lock non re-entrant so that I dont have to deal with multiple layers of unlocking.